lib/determinator/control.rb

```diff
@@ -115,12 +115,41 @@ def choose_target_group(feature, properties)
 
         tg.constraints.reduce(true) do |fit, (scope, *required)|
           present = [*normalised_properties[scope]]
-          fit && (required.flatten & present.flatten).any?
+          fit && (has_any(required, present) || has_any_app_version?(scope, required, present))
         end
       # Must choose target group deterministically, if more than one match
       }.sort_by { |tg| tg.rollout }.last
     end
 
+    def has_any(required, present)
+      (required.flatten & present.flatten).any?
+    end
+
+    def has_any_app_version?(scope, required, present)
+      return false unless scope == 'app_version'
+
+      required_constraints = required.flatten.map {|constraint|
+        begin
+          operator, version = constraint.match(/([<>]=?)(.*)/).captures
+          {:operator => operator, :gem_version => Gem::Version.new(version)}
+        rescue
+          # Ignore malformed constraints
+          next
+        end
+      }.compact
```

Comment:
What's the shape of `required` here? Seems a bit much to flatten and map and compact it

Reply:
Yes sorry, maybe it's not optimal 😬 `required` is as follow `[">=1.2.3","<2.4"]`. I wanted to parse the operator and create the gem version only once before we compare with the versions in `present`.